foreign-chains/solana/sol-prim/src/pda.rs

```diff
@@ -0,0 +1,88 @@
+use digest::Digest;
+use sha2::Sha256;
+
+use crate::{address::Address, consts};
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+#[cfg_attr(feature = "std-error", derive(thiserror::Error))]
+pub enum PdaError {
+	#[cfg_attr(feature = "std-error", error("not a valid point"))]
+	NotAValidPoint,
+	#[cfg_attr(feature = "std-error", error("too many seeds"))]
+	TooManySeeds,
+	#[cfg_attr(feature = "std-error", error("seed too large"))]
+	SeedTooLarge,
+	// TODO: choose a better name
+	#[cfg_attr(feature = "std-error", error("bad luck bumping seed"))]
+	BumpSeedBadLuck,
+}
+
+#[derive(Debug, Clone)]
+pub struct Pda {
+	program_id: Address,
+	hasher: Sha256,
+	seeds_left: u8,
+}
+
+impl Pda {
+	pub fn from_address(program_id: Address) -> Result<Self, PdaError> {
```

Comment:
As we just discussed, this address derivation seems to be an "optimized" version from the `PubKey::try_find_program_address()` used normally in Solana:
https://docs.rs/solana-program/latest/solana_program/pubkey/struct.Pubkey.html#method.try_find_program_address
We really need to make sure that we are not missing anything if we don't use the original code as is. I'd definitely prefer to implement the try_find_program_address as is instead of trying to reinvent the wheel. The upside is to generate deposit addresses slightly more efficiently. The downside is that we have a bug in address derivation that can end up in a catastrophic event.
If we look deep into it and this is a correct simplification of that code and we are sure about that, then it's alright I guess.

Reply:
It looks like a sensible (and pretty simple) optimisation: it just avoids re-hashing all the seeds for every bump. If we're uncertain we can just copy the Solana tests and check it against those.

Reply:
okay, then we go with that :+1:

Reply:
double checked the impl, looks good.